src/mpv.rs

```diff
@@ -96,19 +99,94 @@ unsafe impl GetData for i64 {
     }
 }
 
-unsafe impl SetData for i64 {
-    fn get_format() -> Format {
-        Format::Int64
+#[derive(Debug)]
+pub enum MpvNode<'a> {
+    String(MpvStr<'a>),
+    Flag(bool),
+    Int64(i64),
+    Double(f64),
+    Array(Vec<MpvNode<'a>>),
+    Map(HashMap<MpvStr<'a>, MpvNode<'a>>),
+    None,
+}
+
+impl<'a> TryFrom<libmpv_sys::mpv_node> for MpvNode<'a> {
+    type Error = Error;
+    fn try_from(node: libmpv_sys::mpv_node) -> Result<Self> {
+        Ok(match node.format {
+            mpv_format::Flag => MpvNode::Flag(unsafe { node.u.flag } == 1),
+            mpv_format::Int64 => MpvNode::Int64(unsafe { node.u.int64 }),
+            mpv_format::Double => MpvNode::Double(unsafe { node.u.double_ }),
+            mpv_format::String => {
+                let text = unsafe { mpv_cstr_to_str!(node.u.string) }?;
+                MpvNode::String(MpvStr(text))
+            }
+
+            mpv_format::Array => {
+                let list: libmpv_sys::mpv_node_list = unsafe { *node.u.list };
+                let count = list.num;
+                if count <= 0 {
+                    MpvNode::Array(vec![])
+                } else {
+                    let mut nodes = Vec::with_capacity(count as usize);
+                    for i in 0..count {
+                        let node = unsafe { *list.values.offset(i as isize) };
+                        nodes.push(node.try_into()?);
+                    }
+
+                    MpvNode::Array(nodes)
+                }
+            }
+
+            mpv_format::Map => {
+                let list: libmpv_sys::mpv_node_list = unsafe { *node.u.list };
+                let count = list.num;
+                if count <= 0 {
+                    MpvNode::Map(HashMap::new())
+                } else {
+                    let mut nodes = HashMap::with_capacity(count as usize);
+                    for i in 0..count {
+                        let key = unsafe {
+                            let text = *list.keys.offset(i as isize);
+                            mpv_cstr_to_str!(text)
+                        }?;
+
+                        let node = unsafe { *list.values.offset(i as isize) };
+                        nodes.insert(MpvStr(key), node.try_into()?);
+                    }
+
+                    MpvNode::Map(nodes)
+                }
+            }
+
+            mpv_format::None => MpvNode::None,
+            _ => MpvNode::None,
+        })
     }
 }
 
-unsafe impl GetData for bool {
-    fn get_from_c_void<T, F: FnMut(*mut ctype::c_void) -> Result<T>>(mut fun: F) -> Result<bool> {
+unsafe impl<'a> GetData for MpvNode<'a> {
```

Comment:
Maybe I should `impl` directly on `Vec<MpvNodeMap<'a>>` and `HashMap<MpvStr<'a>, MpvNodeMap<'a>>`? So you could do:

```rust
let cache: HashMap<MpvStr<'_>, MpvNode<'_>> = mpv.get_property("demuxer-cache-state").unwrap();
```

Saves the user from matching the top level enum.

Reply:
Another option would be to implement `TryInto<HashMap<_>>` for `MpvNode` which returns an error if it is not a Map, or even just `Into<HashMap<_>>` that panics.

Then that call would be (assuming type inference groks it):
```rust
let cache: HashMap<MpvStr<'_>, MpvNode<'_>> = mpv.get_property("demuxer-cache-state").unwrap().into();
```

I'm not sure which version I prefer.

Reply:
I think your approach is a more direct translation of the API, which says that [`MPV_FORMAT_NODE_ARRAY` and `_MAP` shouldn't be used directly.](https://github.com/ParadoxSpiral/libmpv-rs/blob/995232c30555a79ad8d1071ef4f1ebdf38841d4a/libmpv-sys/include/client.h#L779)